src/Components/Web.JS/src/Services/RootComponentManager.ts

```diff
@@ -99,11 +120,28 @@ export class RootComponentManager {
     }
   }
 
-  private determinePendingOperation(descriptor: ComponentDescriptor, componentInfo: RootComponentInfo, addIfNewComponent?: boolean): RootComponentOperation | null {
+  private getAutoRenderMode(): 'webassembly' | 'server' | null {
+    // If the WebAssembly runtime has started, we will always use WebAssembly
+    // for auto components. Otherwise, we'll wait to activate root components
+    // until we determine whether the WebAssembly runtime can be loaded quickly.
+
+    if (hasWebAssemblyStarted()) {
+      return 'webassembly';
+    }
+
+    if (this._didWebAssemblyFailToLoadQuickly) {
+      return 'server';
+    }
+
+    return null;
```

Comment:
How will the `null` result be interpreted by the callers here?

Reply:
The `null` indicates that a render mode could not be determined yet. This prevents the component from becoming interactive. At some point in the future (if WebAssembly starts up or fails to load quickly enough), the component will be activated via a call to `updateAllRootComponents()`.